src/http/params.cr

```diff
@@ -78,22 +80,188 @@ module HTTP
+    # params.fetch("non_existent_param") { "default computed value" }    # => "default computed value"
+    # ```
+    def fetch(name)
+      return yield unless has_key?(name)
+      fetch(name)
+    end
+
+    # Appends new value for specified param name. Creates param when there was
+    # no such param.
+    #
+    # ```
+    # params.add("item", "keychain")
+    # params.fetch_all("item")        # => ["pencil", "book", "workbook", "keychain"]
+    # ```
+    def add(name, value)
+      raw_params[name] ||= [] of String
+      raw_params[name] = [] of String if raw_params[name] == [""]
+      raw_params[name] << value
+    end
+
+    # Sets all values for specified param name at once.
+    #
+    # ```
+    # params.set_all("item", ["keychain", "keynote"])
+    # params.fetch_all("item")        # => ["keychain", "keynote"]
+    # ```
+    def set_all(name, values)
+      raw_params[name] = values
+    end
+
+    # Allows to iterate over all name-value pairs.
+    #
+    # ```
+    # params.each do |name, value|
+    #   puts "#{name} => #{value}"
+    # end
+    #
+    # # Outputs:
+    # # email => john@example.org
+    # # item => keychain
+    # # item => keynote
+    # ```
+    def each
+      raw_params.each do |name, values|
+        values.each do |value|
+          yield(name, value)
+        end
+      end
+    end
+
+    # Deletes first value for provided param name. If there are no values left,
+    # deletes param itself. Returns deleted value.
+    #
+    # ```
+    # params.delete("item")     # => "keychain"
+    # params.fetch_all("item")  # => ["keynote"]
+    #
+    # params.delete("item")     # => "keynote"
+    # params["item"]            # KeyError
+    #
+    # params.delete("non_existent_param")  # KeyError
+    # ```
+    def delete(name)
+      value = raw_params[name].shift
+      raw_params.delete(name) if raw_params[name].size == 0
+      value
+    end
+
+    # Deletes all values for provided param name. Returns array of deleted
+    # values.
+    #
+    # ```
+    # params.delete_all("comments")   # => ["hello, world!", ":+1:"]
+    # params.has_key?("comments")     # => false
+    # ```
+    def delete_all(name)
+      raw_params.delete(name)
+    end
+
+    # Serializes to string representation as http url encoded form
+    #
+    # ```
+    # params.to_s     # => "item=keychain&item=keynote&email=john@example.org"
+    # ```
+    def to_s(io)
+      builder = Builder.new(io)
+      each do |name, value|
+        builder.add(name, value)
+      end
+    end
+
     # :nodoc:
     struct Builder
-      def initialize
-        @string = StringIO.new
+      def initialize(@io = StringIO.new)
       end
 
       def add(key, value)
-        @string << '&' unless @string.empty?
-        URI.escape key, @string
-        @string << '='
-        URI.escape value, @string if value
+        @io << '&' unless @io.empty?
```

Comment:
I don't think all IO have this method. The easiest thing would be for the builder to have a `@first` variable, but then it would need to change to a class (no problem with that).